File: IP/ass1/src/synthread2.java
class Synthread2 {
    public static Integer sem1 = 0, sem2 = 0;

    public static void main(String[] args) {
        MyThread2 t = new MyThread2();
        t.start();

        display("ab");
        synchronized (sem1) {
            sem1.notify();
        }

        for(int i = 0; i < 9; i++) {
            synchronized (sem2) {
                try {
                    sem2.wait();
                }
                catch(InterruptedException e) {
                    System.out.println("InterruptedExeption occured, exiting");
                    System.exit(1);
                }
            }
            display("ab");
            synchronized (sem1) {
                sem1.notify();
            }
        }
    }

    static public void display(String str) {
        for(int i = 0; i < str.length(); i++)
            System.out.print(str.charAt(i));
    }
}

class MyThread2 extends Thread {
    private int argument;

    MyThread2() {}

    public void run() {
        for(int i = 0; i < 10; i++) {
            synchronized (Synthread2.sem1) {
                try {
                    Synthread2.sem1.wait();
                }
                catch(InterruptedException e) {
                    System.out.println("InterruptedExeption occured, exiting");
                    System.exit(1);
                }
            }
            Synthread2.display("cd\n");
            synchronized (Synthread2.sem2) {
                Synthread2.sem2.notify();
            }
        }
    }
}

File: IP/ass1/jmr251_1/src/synthread2.c
#include "synthread2.h"

void display(char *str) {
    char *tmp;
    for (tmp=str; *tmp; tmp++) {
        if(!write(1, tmp, 1)) {
            perror("Failed to write to stdout");
            exit(1);
        }
        usleep(100);
    }
}

/* The code to be executed by the original thread. */
void thread1() {
    int i;

    /* This thread is allowed to start without waiting. */
    display("ab");
    sem_post(&sem1);

    for(i=0;i<9;i++) {
        sem_wait(&sem2);
        display("ab");
        sem_post(&sem1);
    }
}

/* The code to be executed by the new thread. */
void *thread2() {
    int i;

    for(i=0;i<10;i++) {
        sem_wait(&sem1);
        display("cd\n");
        sem_post(&sem2);
    }
}

int main() {
    pthread_t thread_id;
    pthread_attr_t attr;

    sem_init(&sem1, 0, 0);
    sem_init(&sem2, 0, 0);

    /* Create the attributes needed for the thread. */
    pthread_attr_init(&attr);

    /* Create the thread and let it execute its code. */
    pthread_create(&thread_id, &attr, thread2, NULL);

    thread1();

    /* Wait for the other thread to be done. */
    pthread_join(thread_id, NULL);

    /* Destroy all created objects. */
    sem_destroy(&sem1);
    sem_destroy(&sem2);
    pthread_attr_destroy(&attr);
    return 0;
}
